File: app/src/main/java/com/reteno/sample/fragments/inbox/InboxMessageAdapter.java
package com.reteno.sample.fragments.inbox;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.reteno.core.domain.model.appinbox.AppInboxMessage;
import com.reteno.sample.databinding.ItemAppinboxBinding;

public class InboxMessageAdapter extends ListAdapter<AppInboxMessage, InboxMessageAdapter.InboxViewHolder> {

    private final InboxItemClick itemClick;

    public InboxMessageAdapter(InboxItemClick itemClick) {
        super(DIFF_CALLBACK);
        this.itemClick = itemClick;
    }

    @NonNull
    @Override
    public InboxViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new InboxViewHolder(
                ItemAppinboxBinding.inflate(
                        LayoutInflater.from(parent.getContext()),
                        parent,
                        false
                ),
                itemClick);
    }

    @Override
    public void onBindViewHolder(InboxViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    static class InboxViewHolder extends RecyclerView.ViewHolder {

        private final ItemAppinboxBinding binding;
        private final InboxItemClick itemClick;

        InboxViewHolder(ItemAppinboxBinding binding, InboxItemClick itemClick) {
            super(binding.getRoot());
            this.binding = binding;
            this.itemClick = itemClick;
        }

        public void bind(AppInboxMessage inbox) {
            binding.tvTitle.setText(inbox.getTitle());
            binding.tvData.setText(inbox.getCreatedDate());
            binding.tvContent.setText(inbox.getContent());
            binding.btnMarkAsOpened.setOnClickListener(v -> {
                itemClick.onOpenedClicked(inbox.getId());
            });
            binding.btnMarkAsOpened.setEnabled(inbox.isNewMessage());

            if (inbox.getLinkUrl() != null && !inbox.getLinkUrl().isEmpty()) {
                binding.cardView.setOnClickListener(v -> {
                    itemClick.onMessageClicked(inbox.getLinkUrl());
                });
            }

            if (inbox.getImageUrl() != null && !inbox.getImageUrl().isEmpty()) {
                Glide.with(binding.imageView)
                        .load(inbox.getImageUrl())
                        .into(binding.imageView);
            }

        }
    }

    public static final DiffUtil.ItemCallback<AppInboxMessage> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<AppInboxMessage>() {
                @Override
                public boolean areItemsTheSame(
                        @NonNull AppInboxMessage oldItem, @NonNull AppInboxMessage newItem) {
                    return oldItem.getId().equals(newItem.getId());
                }

                @Override
                public boolean areContentsTheSame(
                        @NonNull AppInboxMessage oldItem, @NonNull AppInboxMessage newItem) {
                    return oldItem.equals(newItem);
                }
            };

    public interface InboxItemClick {

        void onOpenedClicked(String messageId);

        void onMessageClicked(String url);

    }
}

File: app/src/main/java/com/reteno/sample/fragments/inbox/InboxMessageAdapter.kt
package com.reteno.sample.fragments.inbox

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.reteno.core.domain.model.appinbox.AppInboxMessage
import com.reteno.sample.databinding.ItemAppinboxBinding

class InboxMessageAdapter(private val itemClick: InboxItemClick) :
    ListAdapter<AppInboxMessage, InboxMessageAdapter.InboxViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): InboxViewHolder {
        return InboxViewHolder(
            ItemAppinboxBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ),
            itemClick
        )
    }

    override fun onBindViewHolder(holder: InboxViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    class InboxViewHolder(
        private val binding: ItemAppinboxBinding,
        private val itemClick: InboxItemClick
    ) : RecyclerView.ViewHolder(
        binding.getRoot()
    ) {
        fun bind(inbox: AppInboxMessage) {
            binding.tvTitle.text = inbox.title
            binding.tvData.text = inbox.createdDate
            binding.tvCategory.text = inbox.category
            binding.tvContent.text = inbox.content
            binding.tvStatus.text = inbox.status?.str
            binding.btnMarkAsOpened.setOnClickListener {
                itemClick.onOpenedClicked(inbox.id)
            }
            binding.btnMarkAsOpened.setEnabled(inbox.isNewMessage)
            inbox.linkUrl?.let { link ->
                binding.cardView.setOnClickListener {
                    itemClick.onMessageClicked(link)
                }
                Glide.with(binding.imageView)
                    .load(inbox.imageUrl)
                    .into(binding.imageView)
            }
            binding.imageView.isVisible = inbox.imageUrl != null && inbox.imageUrl!!.isNotEmpty()
            inbox.imageUrl?.let {
                Glide.with(binding.imageView)
                    .load(it)
                    .into(binding.imageView)
            }
            var customData = "empty"
            if (inbox.customData != null && inbox.customData!!.isNotEmpty()) {
                customData = inbox.customData.toString()
            }
            binding.tvCustomData.text = customData
        }
    }

    interface InboxItemClick {
        fun onOpenedClicked(messageId: String)
        fun onMessageClicked(url: String)
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<AppInboxMessage> =
            object : DiffUtil.ItemCallback<AppInboxMessage>() {
                override fun areItemsTheSame(
                    oldItem: AppInboxMessage, newItem: AppInboxMessage
                ): Boolean {
                    return oldItem.id == newItem.id
                }

                override fun areContentsTheSame(
                    oldItem: AppInboxMessage, newItem: AppInboxMessage
                ): Boolean {
                    return oldItem == newItem
                }
            }
    }
}
